### DISCUSSION TITLE
How to declare output parameter in callable type?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#7963] [Q&A]
I'm trying to declare a callable for the following function:
```php
function test(string &$a) : void
{
    $a = 'asdf';
}
```

I've read the discussion from 2018 regarding the callable syntax and the output via parameter reference was mentioned but I think it was overlooked in the initial implementation. Any ideas if it's possible to declare it now?

**Comment by AndrolGenhald** [comment]
Like [this](https://psalm.dev/r/365af05653)? [Here's the documentation](https://psalm.dev/docs/annotating_code/supported_annotations/#param-out-psalm-param-out).

Edit: Or do you mean you want to annotate a callable that has its own `@param-out`? Hmm, not sure if that is possible yet...

**Reply by rzvc** [reply]
No, what I want to declare is something like this:
```php
/** @var callable(&string) $callable */
$callable($a);    // Now $a is 'string'.
```
Sorry for not being clear enough the first time.

**Reply by AndrolGenhald** [reply]
So basically like an annotation to describe a function like:
```<?php
/** @param-out string $bar */
function foo(mixed &$bar): void
{
    ...
}
```

ie `callable(&in<mixed>+out<string>)` would accept `"foo"` as a valid type in that case? (just to throw around a random syntax, not actually suggesting that)

It was clear enough if I've understood it now, it just took me a minute to catch on.

**Reply by rzvc** [reply]
Yep, that's it.

The specification was discussed in #580.
